vshard/router/init.lua

```diff
@@ -873,79 +1022,165 @@ local function router_map_callrw(router, func, args, opts)
-            res, err = future_wait(f, timeout)
-            if res == nil then
+        timeout = deadline - fiber_clock()
+
+        -- Netbox async requests work only with active connections.
+        -- So, first need to wait for the master connection explicitly.
+        local replicasets_to_check = {}
+        group_count = 0
+        for uuid, _ in pairs(grouped_buckets) do
+            group_count = group_count + 1
+            table.insert(replicasets_to_check, replicasets_all[uuid])
+        end
+        timeout, err, err_id = wait_connected_to_masters(
+            replicasets_to_check, timeout)
+        if not timeout then
+            goto fail
+        end
+
+        -- Send ref requests with timeouts to the replicasets.
+        futures = table_new(0, group_count)
+        for id, buckets in pairs(grouped_buckets) do
+            if timeout == nil then
                 err_id = id
                 goto fail
             end
-            -- Map returns true,res or nil,err.
-            res = res:iterator()
-            local count = res:decode_array_header()
-            ok = res:decode()
-            if ok == nil then
-                err = res:decode()
+            local args_ref
+            if replicasets_to_map[id] then
+                -- Replicaset is already referenced on a previous iteration.
+                -- Simply get the moved buckets without double referencing.
+                args_ref = {'storage_moved_buckets', buckets}
+            else
+                args_ref = {'storage_ref_with_lookup', rid, timeout, buckets}
+            end
+            res, err = replicasets_all[id]:callrw('vshard.storage._call',
+                                                  args_ref, opts_async)
+            if res == nil then
                 err_id = id
                 goto fail
             end
-            if count > 1 then
-                map[id] = res:take_array(count - 1)
-            end
-            timeout = deadline - fiber_clock()
+            futures[id] = res
         end
-    else
+
+        -- Wait for the refs to be done and collect moved buckets.
+        bucket_ids = {}
         for id, f in pairs(futures) do
             res, err = future_wait(f, timeout)
+            -- Handle netbox error first.
             if res == nil then
                 err_id = id
                 goto fail
             end
-            -- Map returns true,res or nil,err.
-            ok, res = res[1], res[2]
-            if ok == nil then
-                err = res
+            -- Ref returns nil,err or {rid, moved}.
+            res, err = res[1], res[2]
+            if res == nil then
                 err_id = id
                 goto fail
             end
-            if res ~= nil then
-                -- Store as a table so in future it could be extended for
-                -- multireturn.
-                map[id] = {res}
+            for _, bucket_id in pairs(res.moved) do
+                bucket_reset(router, bucket_id)
```

Comment:
Why don't we return destination, when it's possibl,e alongside with moved bucket ids? This way we won't have to rediscover the bucket from the zero, as we do now. It's not cheap.

AFAICS, it should boost perf of the partial map_callrw during rebalancing

Reply:
No reason really, probably there was no bothering for that. Done now!